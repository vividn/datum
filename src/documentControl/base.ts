import { EitherDocument, EitherPayload } from "./DatumDocument";
import { OutputArgs } from "../input/outputArgs";

export class DocExistsError extends Error {
  data: { existingDocument: EitherDocument; failedPayload: EitherPayload };
  constructor(failedPayload: EitherPayload, existingDocument: EitherDocument) {
    super(
      `conflict: another doc with id "${existingDocument._id}" already exists`,
    );
    this.data = { existingDocument, failedPayload };
    this.name = this.constructor.name;
    Object.setPrototypeOf(this, DocExistsError.prototype);
    Error.captureStackTrace(this, this.constructor);
  }
}

export type BaseDocControlArgs = {
  db: PouchDB.Database<any>;
  outputArgs?: OutputArgs;
};
