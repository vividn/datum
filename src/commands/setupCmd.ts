import { Argv } from "yargs";
import { BaseDatumArgs } from "../input/baseYargs";
import { setupDatumViews } from "../views/setupDatumViews";
import { connectDb } from "../auth/connectDb";
import { Show } from "../output/output";

export const command = "setup";
export const desc = "setup the database for use with datum";

export type SetupCmdArgs = BaseDatumArgs & {
  projectDir?: string;
};

export function builder(yargs: Argv): Argv {
  return yargs.options({
    "project-dir": {
      describe:
        "where to look for additional DatumViews to setup in the database",
      type: "string",
    },
  });
}

export async function setupCmd(args: SetupCmdArgs): Promise<void> {
  const db = connectDb(args);
  const show: Show = args.showAll ? Show.All : args.show ?? Show.None;
  await setupDatumViews({ db, show: show, projectDir: args.projectDir });
}
