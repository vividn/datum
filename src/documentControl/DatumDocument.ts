import {
  isIsoDateOrTime,
  isoDate,
  isoDatetime,
  isoDuration,
} from "../time/timeUtils";
import { WithRequired } from "../utils/utilityTypes";

export type GenericData<T = unknown> = T & {
  [key: string]: any;
};

export type DatumData<T = unknown> = GenericData<T> & {
  occurTime?: isoDatetime | isoDate;
  occurUtcOffset?: number;
  dur?: "start" | "end" | isoDuration;
};

export type OccurredData<T = unknown> = WithRequired<
  DatumData<T>,
  "occurTime" | "occurUtcOffset"
>;

export function isOccurredData(
  data: DatumData | OccurredData
): data is OccurredData {
  const occurTime = data.occurTime;
  return occurTime !== undefined && isIsoDateOrTime(occurTime);
}

export type DatumMetadata = {
  utcOffset?: number;
  createTime?: isoDatetime;
  modifyTime?: isoDatetime; //TODO: turn into an array of times
  idStructure?: string;
  random?: number;
  humanId?: string;
  // [key: string]: any;
};

export type DatumPayload<T = unknown> = {
  _id?: string;
  _rev?: string;
  data: DatumData<T>;
  meta: DatumMetadata;
};

export type DatumDocument<T = unknown> = DatumPayload<T> & {
  _id: string;
  _rev: string;
};

export type DataOnlyPayload<T = unknown> = {
  _id?: string;
  _rev?: string;
} & DatumData<T>;

export type DataOnlyDocument<T = unknown> = DataOnlyPayload<T> & {
  _id: string;
  _rev: string;
};

export type EitherPayload<T = unknown> = DatumPayload<T> | DataOnlyPayload<T>;
export type EitherDocument<T = unknown> =
  | DatumDocument<T>
  | DataOnlyDocument<T>;

export function isDatumDocument(
  doc: DatumDocument | DataOnlyDocument
): doc is DatumDocument {
  return (
    (doc as DatumDocument).data !== undefined &&
    (doc as DatumDocument).meta !== undefined
  );
}

export function isDatumPayload(
  payload: DatumPayload | DataOnlyPayload
): payload is DatumPayload {
  return (
    (payload as DatumPayload).data !== undefined &&
    (payload as DatumPayload).meta !== undefined
  );
}
