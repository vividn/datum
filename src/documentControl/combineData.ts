import { GenericObject } from "../GenericObject";
import { jClone } from "../utils/jClone";
import isPlainObject from "lodash.isplainobject";
import { MergeError } from "../errors";

export type UpdateStrategyNames =
  | "useOld"
  | "useNew"
  | "preferOld"
  | "preferNew"
  | "intersection"
  | "removeConflicting"
  | "xor"
  | "merge"
  | "append"
  | "prepend"
  | "mergeSort"
  | "appendSort";

export const updateStrategies: Record<UpdateStrategyNames, CombiningType> = {
  useOld: { justA: true, justB: false, same: true, conflict: "A" },
  useNew: { justA: false, justB: true, same: true, conflict: "B" },
  preferOld: { justA: true, justB: true, same: true, conflict: "A" },
  preferNew: { justA: true, justB: true, same: true, conflict: "B" },
  intersection: { justA: false, justB: false, same: true, conflict: false },
  removeConflicting: { justA: true, justB: true, same: true, conflict: false },
  xor: { justA: true, justB: true, same: false, conflict: false },
  merge: { justA: true, justB: true, same: true, conflict: "merge" },
  append: { justA: true, justB: true, same: true, conflict: "append" },
  prepend: { justA: true, justB: true, same: true, conflict: "prepend" },
  mergeSort: { justA: true, justB: true, same: true, conflict: "mergeSort" },
  appendSort: { justA: true, justB: true, same: true, conflict: "appendSort" },
};

export type conflictingKeyStrategies =
  | false
  | "A"
  | "B"
  | "merge"
  | "append"
  | "prepend"
  | "mergeSort"
  | "appendSort";

type CombiningType = {
  justA: boolean;
  justB: boolean;
  same: boolean;
  conflict: conflictingKeyStrategies;
};

export function combineData(
  aData: GenericObject,
  bData: GenericObject,
  how: UpdateStrategyNames | CombiningType,
): GenericObject {
  const strategy = typeof how === "string" ? updateStrategies[how] : how;
  const aClone = jClone(aData);
  const bClone = jClone(bData);
  const combined = {} as GenericObject;

  for (const key in aData) {
    const aVal = aClone[key];

    // Just in aData
    if (!(key in bData)) {
      if (strategy.justA) combined[key] = aVal;
      continue;
    }
    const bVal = bClone[key];
    delete bClone[key];

    // If objects in both, then recurse and run strategy on sub object
    if (isPlainObject(aVal) && isPlainObject(bVal)) {
      combined[key] = combineData(aVal, bVal, how);
      continue;
    }

    // In both
    if (aVal === bVal) {
      if (strategy.same) combined[key] = aVal;
      continue;
    }

    // Data is different across a and b
    if (strategy.conflict === "A") {
      combined[key] = aVal;
    } else if (strategy.conflict === "B") {
      combined[key] = bVal;
    } else if (strategy.conflict === false) {
      // pass
    } else if (strategy.conflict === "merge") {
      combined[key] = mergeValues(aVal, bVal, true);
    } else if (strategy.conflict === "append") {
      combined[key] = mergeValues(aVal, bVal, false);
    } else if (strategy.conflict === "prepend") {
      combined[key] = mergeValues(bVal, aVal, false);
    } else if (strategy.conflict === "mergeSort") {
      combined[key] = mergeValues(aVal, bVal, true, true);
    } else if (strategy.conflict === "appendSort") {
      combined[key] = mergeValues(aVal, bVal, false, true);
    } else {
      throw new Error("unknown conflict resolution strategy");
    }
  }

  // just in bData
  if (strategy.justB) {
    for (const bKey in bClone) {
      combined[bKey] = bClone[bKey];
    }
  }

  return combined;
}

const isMergeableValue = (val: any) => {
  return (
    typeof val === "string" ||
    typeof val === "number" ||
    val === null ||
    Array.isArray(val)
  );
};

export const mergeValues = (
  aVal: unknown,
  bVal: unknown,
  unique = true,
  sort = false,
): any => {
  if (bVal === undefined) return aVal;
  if (aVal === undefined) return bVal;

  if (!isMergeableValue(aVal) || !isMergeableValue(bVal)) {
    throw new MergeError(
      "mergeValues is not well defined between these types of values",
    );
  }

  const appended = Array.isArray(aVal)
    ? Array.isArray(bVal)
      ? [...aVal, ...bVal]
      : [...aVal, bVal]
    : Array.isArray(bVal)
      ? [aVal, ...bVal]
      : [aVal, bVal];

  const maybeMerged = unique ? Array.from(new Set(appended)) : appended;
  const maybeSorted = sort ? maybeMerged.sort() : maybeMerged;
  return maybeSorted;
};
