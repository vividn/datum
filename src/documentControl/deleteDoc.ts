import { BaseDocControlArgs } from "./base";
import { isCouchDbError, MyError } from "../errors";
import { Show, showDelete } from "../output/output";

export class NoDocToDeleteError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, NoDocToDeleteError.prototype);
  }
}

type deleteDocType = {
  id: string;
} & BaseDocControlArgs;

export type DeletedDocument = {
  _id: string;
  _rev: string;
  _deleted: true;
};

async function deleteDoc({
  id,
  db,
  show = Show.None,
}: deleteDocType): Promise<DeletedDocument> {
  let existingDoc;
  try {
    existingDoc = await db.get(id);
  } catch (error) {
    if (isCouchDbError(error) && error.error === "not_found") {
      throw new NoDocToDeleteError(`no document exists with id: ${id}`);
    } else {
      throw error;
    }
  }

  const deletedRev = (await db.destroy(id, existingDoc._rev)).rev;
  const deletedDoc = (await db.get(id, { rev: deletedRev })) as DeletedDocument;
  showDelete(existingDoc, show);
  return deletedDoc;
}

export default deleteDoc;
