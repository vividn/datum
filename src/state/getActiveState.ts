import { DateTime } from "luxon";
import { activeStateView } from "../views/datumViews/activeStateView";
import { DatumTime, isDatumTime, now } from "../time/timeUtils";
import { DatumViewMissingError, isCouchDbError } from "../errors";
import { parseTimeStr } from "../time/parseTimeStr";
import { DatumState } from "./normalizeState";

export async function getActiveState(
  db: PouchDB.Database,
  field: string,
  time: DateTime | DatumTime | string = now(),
): Promise<DatumState> {
  const utcTime =
    typeof time === "string"
      ? parseTimeStr({ timeStr: time }).toUTC().toISO()
      : isDatumTime(time)
        ? time.utc
        : time.toUTC().toISO();
  if (utcTime === null) {
    throw new Error("bad time");
  }
  let viewResult;
  try {
    viewResult = await db.query(activeStateView.name, {
      startkey: [field, utcTime],
      descending: true,
      limit: 1,
    });
  } catch (error) {
    //TODO make special error checking missingDatumView
    if (
      isCouchDbError(error) &&
      ["missing", "deleted", "missing_named_view"].includes(error.reason)
    ) {
      throw new DatumViewMissingError(activeStateView.name);
    }
    throw error;
  }

  if (viewResult.rows.length === 0) {
    return null;
  }
  const activeStateRow = viewResult.rows[0];
  if (activeStateRow.key[0] !== field) {
    return null;
  }
  return activeStateRow.value;
}
