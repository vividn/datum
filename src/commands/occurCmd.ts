import { handleTimeArgs, timeArgs, TimeArgs } from "../input/timeArgs";
import { addCmd, AddCmdArgs, newDocArgs } from "./addCmd";
import { EitherDocument } from "../documentControl/DatumDocument";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";
import { dbArgs } from "../input/dbArgs";
import { outputArgs } from "../input/outputArgs";
import { dataArgs } from "../input/dataArgs";
import { fieldArgs } from "../input/fieldArgs";

export const occurArgs = new ArgumentParser({
  add_help: false,
  parents: [fieldArgs, newDocArgs, timeArgs, dataArgs],
});
export const occurCmdArgs = new ArgumentParser({
  description: "record the occurence of something at a single time point",
  prog: "datum occur",
  usage: `%(prog)s <field> [data..]
  %(prog)s --fieldless [data..]
  %(prog)s <field> -K <reqKey1> ... -K <reqKeyN> -k <optKey1>[=defaultVal1] ... -k <optKeyN> <reqVal1> ... <reqValN> [optVal1] ... [optValN] [data..]
`,
  parents: [occurArgs, dbArgs, outputArgs],
});

export type OccurCmdArgs = AddCmdArgs & TimeArgs;

export async function occurCmd(
  args: OccurCmdArgs | string | string[],
  preparsed?: Partial<OccurCmdArgs>,
): Promise<EitherDocument> {
  args = parseIfNeeded(occurCmdArgs, args, preparsed);
  const { time: occurTime } = handleTimeArgs(args);
  if (occurTime !== undefined) {
    args.cmdData ??= {};
    args.cmdData.occurTime = occurTime;
  }

  return await addCmd(args);
}
