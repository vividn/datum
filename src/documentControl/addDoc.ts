import {
  EitherDocument,
  EitherPayload,
  isDatumDocument,
  isDatumPayload,
} from "./DatumDocument";
import { DateTime } from "luxon";
import { IdError, isCouchDbError } from "../errors";
import { jClone } from "../utils/jClone";
import { UpdateStrategyNames } from "./combineData";
import { updateDoc } from "./updateDoc";
import { showCreate, showExists, showFailed } from "../output/output";
import { BaseDocControlArgs, DocExistsError } from "./base";
import isEqual from "lodash.isequal";
import { overwriteDoc } from "./overwriteDoc";
import { deleteDoc } from "./deleteDoc";
import {
  DataOrDesignDocument,
  isViewDocument,
  isViewPayload,
  ViewPayload,
} from "../views/DatumView";
import { assembleId } from "../ids/assembleId";
import { toDatumTime } from "../time/datumTime";

function payloadMatchesDbData(
  payload: EitherPayload,
  existingDoc: EitherDocument,
) {
  const existingWithoutRev = jClone(existingDoc) as EitherPayload;
  delete existingWithoutRev._rev;

  return (
    (isDatumPayload(payload) &&
      isDatumDocument(existingDoc) &&
      isEqual(payload.data, existingDoc.data)) ||
    (isViewPayload(payload) &&
      isViewDocument(existingDoc) &&
      isEqual(payload.views, existingDoc.views)) ||
    (!isDatumPayload(payload) &&
      !isDatumDocument(existingDoc) &&
      isEqual(payload, existingWithoutRev))
  );
}

export type ConflictStrategyNames =
  | UpdateStrategyNames
  | "overwrite"
  | "delete"
  | "fail";

const conflictRecord: Record<ConflictStrategyNames, any> = {
  merge: "",
  useOld: "",
  preferOld: "",
  update: "",
  preferNew: "",
  useNew: "",
  removeConflicting: "",
  xor: "",
  intersection: "",
  append: "",
  prepend: "",
  appendSort: "",
  mergeSort: "",
  overwrite: "",
  delete: "",
  fail: "",
};
export const conflictChoices = Object.keys(conflictRecord);

type addDocType = {
  payload: EitherPayload | ViewPayload;
  conflictStrategy?: ConflictStrategyNames;
} & BaseDocControlArgs;

export async function addDoc({
  db,
  payload,
  conflictStrategy,
  outputArgs = {},
}: addDocType): Promise<DataOrDesignDocument> {
  payload = jClone(payload);
  let id;
  if (payload.meta) {
    const now = toDatumTime(DateTime.local());
    payload.meta.createTime = now;
    payload.meta.modifyTime = now;

    // reassemble id in case it depends upon createTime or modifyTime
    id = assembleId({ payload });

    payload._id = id;

    // Don't care about _rev for adding, will detect and redirect conflicts manually
    delete payload._rev;
  } else {
    id = payload._id;
    if (id === undefined) {
      throw new IdError("id could not be determined");
    }
  }
  try {
    await db.put(payload);
  } catch (error) {
    if (isCouchDbError(error) && error.name !== "conflict") {
      throw error;
    }
    const existingDoc = await db.get(id);

    if (conflictStrategy !== undefined && conflictStrategy !== "fail") {
      if (conflictStrategy === "overwrite") {
        return overwriteDoc({ db, id, payload, outputArgs: outputArgs });
      }
      if (conflictStrategy === "delete") {
        return deleteDoc({ db, id, outputArgs: outputArgs });
      }
      return await updateDoc({
        db,
        id,
        payload,
        updateStrategy: conflictStrategy,
        outputArgs: outputArgs,
      });
    }
    if (payloadMatchesDbData(payload, existingDoc)) {
      showExists(existingDoc, outputArgs);
      return existingDoc;
    } else {
      showFailed(payload, outputArgs);
      showExists(existingDoc, outputArgs);
      throw new DocExistsError(payload, existingDoc);
    }
  }
  const addedDoc = await db.get(id);
  showCreate(addedDoc, outputArgs);
  return addedDoc;
}
