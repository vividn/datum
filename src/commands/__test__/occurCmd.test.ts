import {
  deterministicHumanIds,
  popNow,
  pushNow,
  restoreNow,
  setNow,
  testDbLifecycle,
} from "../../__test__/test-utils";
import { DatumDocument } from "../../documentControl/DatumDocument";
import { occurCmd } from "../occurCmd";
import { setupCmd } from "../setupCmd";
import { getActiveState } from "../../state/getActiveState";

describe("occurCmd", () => {
  const dbName = "occur_cmd_test";
  const db = testDbLifecycle(dbName);

  beforeEach(async () => {
    await setupCmd("");
  });
  afterEach(() => {
    restoreNow();
  });

  it("creates a document with an occurTime", async () => {
    const now = "2023-08-05T16:00:00.000Z";
    setNow(now);

    const docCountBefore = (await db.info()).doc_count;
    const doc = await occurCmd("field");

    await db.info().then((info) => {
      expect(info.doc_count).toEqual(docCountBefore + 1);
    });
    expect(doc.data.field).toEqual("field");
    expect(doc.data.occurTime.utc).toEqual(now);

    expect(doc._id).toEqual(`field:${now}`);
    const dbDoc = await db.get(doc._id);
    expect(dbDoc).toEqual(doc);
  });

  it("stores the occurTime in the data", async () => {
    const newDoc = (await occurCmd(
      "event -d 2021-08-23 -t 12 -z 0",
    )) as DatumDocument;
    expect(newDoc.data).toHaveProperty(
      "occurTime.utc",
      "2021-08-23T12:00:00.000Z",
    );
    expect(newDoc.meta).not.toHaveProperty("occurTime");
  });

  it("stores the occurTime in DataOnly docs", async () => {
    const newDoc = (await occurCmd(
      "event -M -d 2021-08-23 -t 12 -z 0",
    )) as DatumDocument;
    expect(newDoc).toHaveProperty("occurTime.utc", "2021-08-23T12:00:00.000Z");
    expect(newDoc).toHaveProperty("occurTime.o", 0);
  });

  it("stores offset", async () => {
    const newDoc = (await occurCmd(
      "event -d 2021-08-23 -t 12 -z +3",
    )) as DatumDocument;
    expect(newDoc.data).toMatchObject({
      occurTime: {
        o: 3,
        tz: "UTC+3",
      },
    });
  });

  it("stores offset from an IANA timezone", async () => {
    const newDoc = (await occurCmd(
      "event -d 2023-12-04 -t 12 -z Europe/Berlin",
    )) as DatumDocument;
    expect(newDoc.data).toMatchObject({
      occurTime: {
        o: 1,
        tz: "Europe/Berlin",
      },
    });
  });

  it("records lastState as null if the active state is null", async () => {
    expect(await getActiveState(db, "event")).toBe(null);
    const newDoc = await occurCmd("event");
    expect(newDoc.data).toMatchObject({
      lastState: null,
    });
    // Even 1 occur document is enough to switch active state to false from null
    expect(await getActiveState(db, "event")).toBe(false);

    const newDoc2 = await occurCmd("event");
    expect([false, undefined]).toContainEqual(newDoc2.data.lastState);
  });

  it("handles negative number time arguments correctly", async () => {
    pushNow("2024-02-23,16:00");
    const newDoc = await occurCmd("event -d -1d -t -1h");
    expect(newDoc.data.occurTime.utc).toEqual("2024-02-22T15:00:00.000Z");
    const newDoc2 = await occurCmd("event -t -20");
    expect(newDoc2.data.occurTime.utc).toEqual("2024-02-23T15:40:00.000Z");
    popNow();
  });

  it("can occur on a full day", async () => {
    pushNow("2024-04-23,13:30");
    const newDoc = await occurCmd("event -D");
    expect(newDoc.data.occurTime).toMatchObject({
      utc: "2024-04-23",
    });

    const newDoc2 = await occurCmd("happening --full-day");
    expect(newDoc2.data.occurTime).toMatchObject({
      utc: "2024-04-23",
    });
  });

  describe("change command", () => {
    deterministicHumanIds();

    beforeEach(async () => {
      setNow("2023-12-21 14:00");
    });
    afterAll(() => {
      restoreNow();
    });

    it("can become a start command by having start as a trailing word", async () => {
      expect(
        await occurCmd("field -K req1 -k opt1 reqVal optVal start 30min"),
      ).toMatchSnapshot({
        _rev: expect.any(String),
      });
    });

    it("can become an end command by having start as a trailing word", async () => {
      expect(
        await occurCmd("field -K req1 -k opt1 reqVal optVal end 30min"),
      ).toMatchSnapshot({
        _rev: expect.any(String),
      });
    });

    it("can become a switch command by having start as a trailing word", async () => {
      expect(
        await occurCmd(
          "field -K req1 -k opt1 reqVal optVal switch stateName 5m30s",
        ),
      ).toMatchSnapshot({
        _rev: expect.any(String),
      });
    });
  });
});
