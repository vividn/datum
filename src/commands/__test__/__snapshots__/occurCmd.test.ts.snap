// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`occurCmd change command can become a start command by having start as a trailing word 1`] = `
Object {
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": Object {
    "dur": "PT30M",
    "field": "field",
    "lastState": null,
    "occurTime": Object {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": true,
  },
  "meta": Object {
    "createTime": Object {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%field%:%occurTime%",
    "modifyTime": Object {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`occurCmd change command can become a switch command by having start as a trailing word 1`] = `
Object {
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": Object {
    "dur": "PT5M30S",
    "field": "field",
    "lastState": null,
    "occurTime": Object {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": "stateName",
  },
  "meta": Object {
    "createTime": Object {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%field%:%occurTime%",
    "modifyTime": Object {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`occurCmd change command can become an end command by having start as a trailing word 1`] = `
Object {
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": Object {
    "dur": "PT30M",
    "field": "field",
    "lastState": null,
    "occurTime": Object {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": false,
  },
  "meta": Object {
    "createTime": Object {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%field%:%occurTime%",
    "modifyTime": Object {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;
